src/mesh/voxel_cloud.rs

```diff
@@ -0,0 +1,456 @@
+                                // and set_voxel_at_absolute_coords a voxel containing the point to be on
+                                voxel_cloud.set_voxel_at_cartesian_coords(true, &cartesian);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        voxel_cloud
+    }
+
+    /// For each existing voxel turn on all neighbor voxels to grow (offset) the
+    /// volumes stored in the voxel cloud.
+    pub fn grow(&self) -> Self {
+        // If the voxels in the existing voxel cloud reach the boundaries of the
+        // block, it's needed to grow the block by 1 in each direction.
+        let new_block_start = self.block_start - Vector3::new(-1, -1, -1);
+
+        let new_block_dimensions = self.block_dimensions + Vector3::new(2, 2, 2);
+        let mut voxel_cloud = VoxelCloud::new(
+            &new_block_start,
+            &new_block_dimensions,
+            &self.voxel_dimensions,
+        );
+
+        // Iterate through the existing voxel cloud.
+        for z in 0..self.block_dimensions.z {
+            for y in 0..self.block_dimensions.y {
+                for x in 0..self.block_dimensions.x {
+                    let voxel_coords = Point3::new(cast_i32(x), cast_i32(y), cast_i32(z));
+                    let voxel_state = self.voxel_at_relative_coords(&voxel_coords);
+                    // if the voxel is on
+                    if let Some(true) = voxel_state {
+                        // set it to be on also in the new voxel cloud
+                        // (everything is shifted by 1, 1, 1 because the start
+                        // is shifted -1, -1, -1) as well as all its neighbors
+                        let neighbor_offsets = [
+                            Vector3::new(1, 1, 1), //self
+                            Vector3::new(0, 1, 1), //neighbors
+                            Vector3::new(2, 1, 1),
+                            Vector3::new(1, 0, 1),
+                            Vector3::new(1, 2, 1),
+                            Vector3::new(1, 1, 0),
+                            Vector3::new(1, 1, 2),
+                        ];
+
+                        for neighbor_offset in &neighbor_offsets {
+                            voxel_cloud.set_voxel_at_relative_coords(
+                                true,
+                                &(voxel_coords + neighbor_offset),
+                            );
+                        }
+                    }
+                }
+            }
+        }
+
+        voxel_cloud
+    }
+
+    /// Gets the state of a voxel defined in voxel coordinates relative to the
+    /// voxel block start.
+    pub fn voxel_at_relative_coords(&self, relative_coords: &Point3<i32>) -> Option<bool> {
+        match self.relative_three_dimensional_coordinate_to_one_dimensional(relative_coords) {
+            Some(index) => Some(self.voxel_map[index]),
+            None => None,
+        }
```

Comment:
is there a special implementation of `map` for `Option`?

Reply:
Yes. `Option` has many combinators. They are hardcoded (i.e. not generic) in rust, because unlike haskell, rust doesn't have higher-kinded types and monads, but they are present.

https://doc.rust-lang.org/std/option/enum.Option.html#method.map